wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/DeferredCommand.java

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package edu.wpi.first.wpilibj2.command;
+
+import static edu.wpi.first.util.ErrorMessages.requireNonNullParam;
+
+import edu.wpi.first.util.sendable.SendableBuilder;
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * Runs the given command when this command is initialized, and ends when it ends. Useful for
+ * performing runtime tasks before creating a new command. If this command is interrupted, it will
+ * cancel the command.
+ *
+ * <p>This class is provided by the NewCommands VendorDep
+ */
+public class DeferredCommand extends Command {
+  private final Supplier<Command> m_supplier;
+  private Command m_command;
+
+  /**
+   * Creates a new DeferredCommand that runs the supplied command when initialized, and ends when it
+   * ends. Useful for lazily creating commands at runtime. The {@link Supplier} will be called each
+   * time this command is initialized. The Supplier <i>must</i> create a new Command each call.
+   *
+   * @param supplier The command supplier
+   * @param requirements The command requirements. This is a {@link Set} to prevent accidental
+   *     omission of command requirements.
+   */
+  public DeferredCommand(Supplier<Command> supplier, Set<Subsystem> requirements) {
+    m_supplier = requireNonNullParam(supplier, "supplier", "DeferredCommand");
+    addRequirements(requirements.toArray(new Subsystem[0]));
+    resetInternalCommand();
+  }
+
+  @Override
+  public void initialize() {
+    var cmd = m_supplier.get();
+    if (cmd != null) {
+      m_command = cmd;
+    }
+    CommandScheduler.getInstance().registerComposedCommands(m_command);
+    m_command.initialize();
+  }
+
+  @Override
+  public void execute() {
+    m_command.execute();
+  }
+
+  @Override
+  public boolean isFinished() {
+    return m_command.isFinished();
+  }
+
+  @Override
+  public void end(boolean interrupted) {
+    m_command.end(interrupted);
+    resetInternalCommand();
+  }
+
+  @Override
+  public void initSendable(SendableBuilder builder) {
+    super.initSendable(builder);
+    builder.addStringProperty(
+        "deferred", () -> m_command == null ? "null" : m_command.getName(), null);
+  }
+
+  private void resetInternalCommand() {
+    m_command = new PrintCommand("[DeferredCommand] Supplied command was null!");
```

Comment:
Would it be better to not recreate this "null" command, rather storing it as a (static?) final field that is marked as composed, and moving the `registerComposedCommands` call in `initialize` into the `if` statement?

Reply:
I don't see the need to mark it as composed- it's an implementation detail of the command and won't ever exit the command.

I don't think it should be static- this sort of violates not having multiple compositions use a single command. It doesn't matter for this specific use case (PrintCommand not having internal state) so maybe it's fine?